### PR TITLE
Adding the Elo Worth for Complexity

### PR DESCRIPTION
**Pull request by FauziAkram**
Adding a roughly estimated Elo Worth for Complexity Formula

    int complexity =   9 * pe->passed_count()       // (~  3 Elo)
                    + 12 * pos.count<PAWN>()        // (~ 40 Elo)
                    +  9 * outflanking              // (~  2 Elo)
                    + 21 * pawnsOnBothFlanks        // (~0.5 Elo)
                    + 24 * infiltration             // (~  2 Elo)
                    + 51 * !pos.non_pawn_material() // (~0.5 Elo)
                    - 43 * almostUnwinnable         // (~0.5 Elo)
                    -110 ;                          // (~ 20 Elo)

### CODE REVIEW

**Comment by vondele** [comment]
I think this estimate is probably not so useful, as the constant adjustment (-110) should probably be 'split' over the individual terms.  Just removing `12 * pos.count<PAWN>()` indeed costs 40 Elo, but if one adjusts `-110` at the same time by something like 12 * averageNumberOfPawns, the result would be very different.

**Comment by Lolligerhans** [comment]
This is a general problem. It applies to king danger estimates present in master in just the same way.

It may help interpretation to know the total elo worth of the entire initiative adjustment + the impact of each indicidual term + standard deviation (SD) (denoted `o`). Means and SD of the complexity terms; top down:

During openings`*`:

```
    [0    ] Samples 35017625 Mean 3.43635 o 5.81412
    [1    ] Samples 35017625 Mean 168.874 o 18.2372
    [2    ] Samples 35017625 Mean -53.4492 o 11.9983
    [3    ] Samples 35017625 Mean 20.9998 o 0.0570012
    [4    ] Samples 35017625 Mean 0.000651786 o 0.12507
    [5    ] Samples 35017625 Mean 0 o 0
    [6    ] Samples 35017625 Mean -0.000316812 o 0.116717
    [total] Samples 35017625 Mean 29.8618 o 21.2094
```

Generally`**`:
```
    [0    ] Samples 28813990 Mean 8.81255 o 9.88492
    [1    ] Samples 28813990 Mean 97.9207 o 52.7336
    [2    ] Samples 28813990 Mean -30.1999 o 32.4131
    [3    ] Samples 28813990 Mean 16.9729 o 8.2675
    [4    ] Samples 28813990 Mean 3.724 o 8.68952
    [5    ] Samples 28813990 Mean 1.434 o 8.43075
    [6    ] Samples 28813990 Mean -3.3192 o 11.4764
    [total] Samples 28813990 Mean -14.6549 o 47.8676
```

Note how, in some sense, the standard deviation is more important than the average bosnus: If` SD==0`, then the term is not actually used to differentiate between positions, irrespective ot its mean. It is entirely counteracted by the tuned offset `-110`.

Elo worth of the offset parameter is pointless in any case imo.

---

- `*` `./stockfish bench 1 1 12 "closedpos_deep_short.epd";`  In other tests appeared representative for the general, opening-middlegame-ish position. (Contains 999 positions)
- `**` `./stockfish bench 1 1 12 "pos1000.fen";`  I use this because I don't have an endgame-specialized file :) (Contains 1000 positions)

**Comment by ddugovic** [comment]
If you knew covariance between the terms and some of these truly are ~0.5 Elo, you'd probably be able to figure out how to simplify away a term or two by now.

**Comment by Lolligerhans** [comment]
Additionally, we obtain a hint to which bonuses are not yet optimally refined.

### Example

The `outflanking` term has much larger SD than the `passed_count` term (more potential to distinguish positions from each other). Yet, it's elo worth is similar or smaller (granted, the larger mean probably biased more heavily upon removal-testing). This may indicate the the `outflanking` bonus is not very useful in general, or, has not reached it's full potential yet.
From here we develop ideas.
For example, one could identify the edge cases of maximum impact and see if some detrimental edge cases, preventing us from using larger multipliers, can be removed.
- Intuitively, very small values of outflanking (say, below -2 or -3) are not very telling in terms of kings outflanking each other. Yet the scale can reach -7 comfortably. If the intuition is valid, we may be able to use larger multipliers if this large impact edge case is removed. So cap outflanking at -2/-3, maybe increase multiplier.
- The otherextreme is large outflanking. Likewise, the maximum outflanking of +14 may not be much different to +2 to +5 in terms of intuitive relevane. So we try clamping outflanking in [-3, 5].
- Finally, combining outflanking with `pawnsOnBothFlanks` sounds like a pawn race to queening, in which case we may want to keep every bit of value the passer code gave us. Try increasing initiative in this case.

### tl;dr
I argue for adding elo worth together with mean and SD because in combination they do become useful.

**Comment by Lolligerhans** [comment]
@ddugovic Imho, when a larger correlation between terms is found, the instinct should be to reduce their correlation by refining them, not to maximize it by merging them into one.

From Leela and nnue we now know for a fact that we can expect to beneficially use orders of magnitude more compute for static, as long as the terms remain relevant.

**Comment by vondele** [comment]
I'll close this issue without merging, generally I find it is not bad to document importance of terms in search and eval, but this is a bit too fine-grained for my taste, and not fully informative.
